src/index.js

```diff
@@ -0,0 +1,205 @@
+
+const Timestamp = MongoDB.Timestamp;
+const defaultKeyPrefix = 'mongoOplogReader';
+
+class MongoOplogReader extends EventEmitter {
+
+  /**
+   * Instantiate a new MongoOplogReader
+   * @param options
+   *  - connectionStrings {String[]} a list of mongodb connection strings
+   *  - redisClient {Object}
+   *  - keyPrefix {String} the redis key prefix (default: mongoOplogReader)
+   */
+  constructor(options) {
+    super();
+    if (!Array.isArray(options.connectionStrings) || !options.connectionStrings.length) {
+      throw new Error('connectionStrings is required.');
+    }
+    if (!options.redisClient || !options.redisClient.sadd) {
+      throw new Error('redisClient must support GET,SET,SADD,SCARD commands.');
+    }
+    this.connectionStrings = options.connectionStrings;
+    this.redisClient = options.redisClient;
+    this.oplogs = {};
+    this.replSets = {};
+    this.keyPrefix = options.keyPrefix || defaultKeyPrefix;
+  }
+
+  /**
+   * Begin tailing all oplogs
+   * @param options
+   *  - since {BsonTimestamp|Int} - the timestamp to start tailing the oplogs from
+   */
+  tail(options) {
+    if (options && options.since) {
+      return this.connectionStrings.forEach(connStr => this.tailHost(connStr, options));
+    }
+    this.getLastOpTimestamp()
+      .then(ts => {
+        const opts = Object.assign({}, options, { since: ts });
+        this.connectionStrings.forEach(connStr => this.tailHost(connStr, opts));
+      })
+      .catch(console.log);
+  }
+
+  getLastOpTimeKey(replSetName) {
+    return `${this.keyPrefix}:${replSetName}:lastOpTime`;
+  }
+
+  setLastOpTimestamp(replSetName, data) {
+    const key = this.getLastOpTimeKey(replSetName);
+    const unixTime = data.ts.high_;
+    const opSeq = data.ts.low_;
+    const val = `${unixTime}|${opSeq}`;
+    this.redisClient.set(key, val);
+  }
+
+  getLastOpTimestamp(replSetName) {
+    const key = this.getLastOpTimeKey(replSetName);
+    return new Promise((resolve, reject) => {
+      this.redisClient.get(key, (err, val) => {
+        if (err) return reject(err);
+        if (!val) return resolve();
+        const [unixTimestamp, opSeq] = val.split('|').map(n => Number(n));
+        const ts = Timestamp(opSeq, unixTimestamp);
+        resolve(ts);
+      })
+    });
+  }
+
+  setConnectionStrings(newConnStrings) {
+    // destroy old/unneeded oplog connections
+    Object.keys(this.oplogs).forEach(connStr => {
+      if (!newConnStrings.includes(connStr)) {
+        this.oplogs[connStr].destroy()
+          .then(() => {
+            console.log('removing oplog:', connStr);
+            delete this.oplogs[connStr];
+          });
+      }
+    });
+    newConnStrings.forEach(newConnStr => {
+      const exists = Object.keys(this.oplogs).includes(newConnStr);
+      if (!exists) {
+        // start reading the oplog of a new shard
+        console.log('new oplog:', newConnStr);
+        // TODO: do we need to start the tail at a time in the past?
```

Comment:
I think it's ok to start tailing "now" in the case that a new shard is added. Thoughts?

Reply:
Im wondering if we should do now-1 hour, for safety reasons. What if there's a write just a millisecond before now?

Reply:
I'll start tailing from the beginning of the oplog if this replset has no `lastOpTime` in redis.